config/config.go

```diff
@@ -166,14 +223,327 @@ func (cs configs) ResolveConfig(f func(configs []interface{}) error) error {
+		return func(*stscreds.AssumeRoleOptions) {}, false, nil
+	}
+
+	return o.AssumeRoleCredentialOptions, true, nil
+}
+
+// WithAssumeRoleCredentialOptions  is a helper function to construct functional options
+// that sets a function to use stscreds.AssumeRoleOptions on config's LoadOptions
+func WithAssumeRoleCredentialOptions(v func(*stscreds.AssumeRoleOptions)) func(*LoadOptions) error {
+	return func(o *LoadOptions) error {
+		o.AssumeRoleCredentialOptions = v
+		return nil
+	}
+}
+
+func (o LoadOptions) getHTTPClient(ctx context.Context) (HTTPClient, bool, error) {
+	if o.HTTPClient == nil {
+		return nil, false, nil
+	}
+
+	return o.HTTPClient, true, nil
+}
+
+// WithHTTPClient is a helper function to construct functional options
+// that sets HTTPClient on LoadOptions
+func WithHTTPClient(v HTTPClient) func(*LoadOptions) error {
+	return func(o *LoadOptions) error {
+		o.HTTPClient = v
+		return nil
+	}
+}
+
+func (o LoadOptions) getAPIOptions(ctx context.Context) ([]func(*middleware.Stack) error, bool, error) {
+	if o.APIOptions == nil {
+		return []func(stack *middleware.Stack) error{}, false, nil
+	}
+
+	return o.APIOptions, true, nil
+}
+
+// WithAPIOptions is a helper function to construct functional options
+// that sets APIOptions on LoadOptions
+func WithAPIOptions(v func(*middleware.Stack) error) func(*LoadOptions) error {
+	return func(o *LoadOptions) error {
+		o.APIOptions = append(o.APIOptions, v)
+		return nil
+	}
+}
+
+func (o LoadOptions) getRetryer(ctx context.Context) (aws.Retryer, bool, error) {
+	if o.Retryer == nil {
+		return nil, false, nil
+	}
+
+	return o.Retryer, true, nil
+}
+
+// WithRetryer is a helper function to construct functional options
+// that sets Retryer on LoadOptions
+func WithRetryer(v aws.Retryer) func(*LoadOptions) error {
+	return func(o *LoadOptions) error {
+		o.Retryer = v
+		return nil
+	}
+}
+
+func (o LoadOptions) getEndpointResolver(ctx context.Context) (aws.EndpointResolver, bool, error) {
+	if o.EndpointResolver == nil {
+		return nil, false, nil
+	}
+
+	return o.EndpointResolver, true, nil
+}
+
+// WithEndpointResolver is a helper function to construct functional options
+// that sets endpoint resolver on LoadOptions
+func WithEndpointResolver(v aws.EndpointResolver) func(*LoadOptions) error {
+	return func(o *LoadOptions) error {
+		o.EndpointResolver = v
+		return nil
+	}
+}
+
+func (o LoadOptions) getLogger(ctx context.Context) (logging.Logger, bool, error) {
+	if o.Logger == nil {
+		return nil, false, nil
+	}
+
+	return o.Logger, true, nil
+}
+
+// WithLogger is a helper function to construct functional options
+// that sets Logger on LoadOptions
+func WithLogger(v logging.Logger) func(*LoadOptions) error {
+	return func(o *LoadOptions) error {
+		o.Logger = v
+		return nil
+	}
+}
+
+func (o LoadOptions) getClientLogMode(ctx context.Context) (aws.ClientLogMode, bool, error) {
+	if o.ClientLogMode == 0 {
+		return 0, false, nil
+	}
+
+	return o.ClientLogMode, true, nil
+}
```

Comment:
This would make it impossible for user to set ClientLogMode to a zero value via a functional option.  I think we should take a step back and reconsider the value of these `bool` vs value parameters for the option.

Arguably in this case the bool doesn't really have a value, and actually creates complexity.  Without the bool, the value of the `ClientLogMode` is always returned.

This is the reason for the LoadDefaultConfig function initializing the `aws.Config` embedded field of `LoadOptions` **before** processing the functional options provided by the caller, this ensures that the SDK has initialized all config defaults appropriately, before applications functional options are applied.